### ISSUE TITLE
Zig stage2 erroneous backwards branch evaluation/segfault with recursive function

### ISSUE THREAD

**Issue by jwhear** [ziglang/zig#13013]
### Zig Version

0.10.0-dev.4198+5e0d8a435

### Steps to Reproduce

Run `zig test` on the following code (heavily reduced to minimize surface area):
```zig
const std = @import("std");
const swap = std.mem.swap;

pub fn stablePartition(comptime T: type, slice: []T, predicate: fn(T) bool) i8 {
    if (slice.len == 0)
        return 0;
    _ = stablePartition(T, slice[1..], predicate);
    return 1;
}
test "stablePartition" {
    var items = [_]i8{0,1,2,3,4,5,6};
    _=stablePartition(i8, &items, isOdd);
}

fn isOdd(val: i8) bool {
    return val & 1 == 1;
}
```

By default this produces the following error:
```
min.zig:7:24: error: evaluation exceeded 1000 backwards branches
    _ = stablePartition(T, slice[1..], predicate);
        ~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~
min.zig:7:24: note: use @setEvalBranchQuota() to raise the branch limit from 1000
min.zig:7:24: note: called from here (999 times)
min.zig:14:22: note: called from here
    _=stablePartition(i8, &items, isOdd);
      ~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~
```

Using `@setEvalBranchQuota(5000)` as suggested causes the compiler to crash on my machine.  Note that this is during compilation, not runtime.

### Expected Behavior

Running `zig test -fstage1 min.zig` compiles and runs successfully with all tests passing.

### Actual Behavior

Backwards branches error or segfault if the branch quota is increased.  There appears to be some kind of infinite recursion, perhaps in type inference/resolution.

**Comment by Vexu** [comment]
Duplicate of #12973

**Comment by topolarity** [comment]
I'm not sure this is a duplicate of #12973, since `stablePartition` should only be instantiated once.

I think we're missing "error: parameter of type 'fn (i8) bool' must be declared comptime"  

@jwhear The example works if you update the function pointer for stage2:
```zig
pub fn stablePartition(comptime T: type, slice: []T, predicate: std.meta.FnPtr(fn(T) bool)) i8
```
`std.meta.FnPtr` returns `fn(...)` for stage1, and `*const fn(...)` for stage2

**Comment by jwhear** [comment]
Ah, good catch.  I distantly remember reading this in the transition guide a while ago, but had forgotten all about it.  Using `FnPtr` should unblock my code, though the compiler bug remains.

**Comment by Vexu** [comment]
This is then related to #13022 and probably caused by 8e4d0ae4f5c50621a402192f2ff7f9b156030257

**Comment by perillo** [comment]
@Vexu with `0.12.0-dev.2341+92211135` the example compiles correctly, without the need to use `@setEvalBranchQuota`.

**Comment by Vexu** [comment]
The original example works when converting the test into an exported function but the version with a test hits an assertion failure.

**Comment by perillo** [comment]
> The original example works when converting the test into an exported function but the version with a test hits an assertion failure.

Strange, since I ran `zig test` on the original example, getting
```console
All 1 tests passed.
```

I'm on Linux.

**Comment by Vexu** [comment]
My mistake, I wasn't using master branch. On a positive note, I figured out why #18712 isn't passing CI.

It does compile now but probably shouldn't as noted by topolarity above.

**Comment by mlugg** [comment]
The original repro now compiles fine, and it's correct for it to do so, because the `comptime` annotation isn't required on generic parameters.

I *believe* the original behavior here was because the self-hosted compiler used to upgrade calls to inline calls when a generic instantiation resulted in a parameter not marked `comptime` having a comptime-only type. Regardless, the current behavior is fine; this just needs a test.
